### ISSUE TITLE
openssl genpkey -algorithm rsa-pss generates invalid keys

### ISSUE THREAD

**Issue by bdhess** [openssl/openssl#20540]
Example:

```
openssl genpkey -algorithm rsa-pss -pkeyopt rsa_keygen_bits:2048 | openssl asn1parse
    0:d=0  hl=4 l=1211 cons: SEQUENCE          
    4:d=1  hl=2 l=   1 prim: INTEGER           :00
    7:d=1  hl=2 l=  11 cons: SEQUENCE          
    9:d=2  hl=2 l=   9 prim: OBJECT            :rsassaPss
   20:d=1  hl=4 l=1191 prim: OCTET STRING      [HEX DUMP]:308204A302010002820101
   (snip)
```

Per [RFC 8017 Appendix A.2.3](https://www.rfc-editor.org/rfc/rfc8017#appendix-A.2.3), `id-RSASSA-PSS` is a scheme identifier.  (Not a key identifier.)

Per [RFC 8017 Appendix A.1](https://www.rfc-editor.org/rfc/rfc8017#appendix-A.1) the object identifier `rsaEncryption` identifies RSA public and private keys.

**Comment by t8m** [comment]
I do not think this is a bug. The rsassaPss identifier in certificates can identify keys which are restricted to be used only for RSA-PSS signatures. I do not see a reason why it could not be extended to private keys.

You can always use `-algorithm rsa` to generate a regular RSA key so there is no point in changing this.

**Comment by t8m** [comment]
Then you just use `-algorithm rsa`. If this was the only type of RSA keys generated, then I'd agree we have a problem and must fix it. But as it is not, I do not see a reason to change/block this. However, perhaps we should at least document this incompatibility problem.

**Comment by gingerbeard-man** [comment]
Question:  How do the other libraries ensure that an RSA-PSS key is not misused to create a regular (non-PSS) RSA signature?  I quite like the fact that `openssl` simply says "no".

As PKCS#11 only knows one type of RSA key we have the same basic problem with our "PSS only" CAs.  The HSM could happily create either type of signature so how do we convince the auditor that it will only ever do RSA-PSS for a particular key..?

**Comment by bdhess** [comment]
With PKCS#11 you could set CKA_ALLOWED_MECHANISMS to CKM_RSA_PKCS_PSS.

**Comment by bdhess** [comment]
@t8m a documentation change would be good. Also some way to convert the key to standard if your user wants to port it out.  Maybe allow it to be converted with `pkcs8 -topk8` (which presently doesn’t change the OID)?

**Comment by t8m** [comment]
You can use openssl rsa -in pss.key -out rsa.key -traditional to convert it to the old PKCS#1 format which does not contain any algorithm identifier.
